### PR TITLE
Feature: add ban cheat versions

### PR DESCRIPTION
**Pull request by BloodWod**

### CODE REVIEW

File: src/game/server/gamecontext.cpp [hunk 1]
```diff
@@ -1686,6 +1686,15 @@ void CGameContext::OnMessage(int MsgID, CUnpacker *pUnpacker, int ClientID)
 	void *pRawMsg = m_NetObjHandler.SecureUnpackMsg(MsgID, pUnpacker);
 	CPlayer *pPlayer = m_apPlayers[ClientID];
 	
+	if (pPlayer && MsgID == (NETMSGTYPE_CL_CALLVOTE + 1)) 
+	{
+        int Version = pUnpacker->GetInt();
```
Comment (bretonium) [inline src/game/server/gamecontext.cpp:1691]: please change the whitespaces to tabs

**Comment by bretonium** [comment]
This is great, but when i try to compile it i get
```
[29/77] #1 c src/base/system.c
src/base/system.c: In function ‘str_in_list’:
src/base/system.c:1604:27: warning: implicit declaration of function ‘str_token_get’ [-Wimplicit-function-declaration]
  while(notfound && (tok = str_token_get(tok, delim, &len)))
                           ^~~~~~~~~~~~~
src/base/system.c:1604:25: warning: assignment makes pointer from integer without a cast [-Wint-conversion]
  while(notfound && (tok = str_token_get(tok, delim, &len)))
                         ^
```

Could you please fix it?

**Comment by bretonium** [comment]
Actually, this is not even a warning, after some time i get

```
objs/base/system_sql.o: In function `str_in_list':
system.c:(.text+0x16dc): undefined reference to `str_token_get'
collect2: error: ld returned 1 exit status
bam: 'bin/server_sql' error 256
bam: 'bin/server_sql' removed because job updated it even it failed.
bam: error: a build step failed
```

**Comment by bretonium** [comment]
```
[ 1/31] #1 c++ src/engine/server/server.cpp
src/engine/server/server.cpp: In static member function ‘static bool CServer::ConStatusExtended(IConsole::IResult*, void*)’:
src/engine/server/server.cpp:2466:27: error: ‘class IGameServer’ has no member named ‘GetClientVersion’; did you mean ‘NetVersion’?
      pThis->GameServer()->GetClientVersion(i);
                           ^~~~~~~~~~~~~~~~
src/engine/server/server.cpp:2467:5: error: expected primary-expression before ‘)’ token
     );
     ^
bam: 'objs/engine/server/server_sql.o' error 256
```

**Comment by bretonium** [comment]
compile error again:
```
[1/8] #1 c++ src/engine/server/server.cpp
src/engine/server/server.cpp: In static member function ‘static bool CServer::ConStatusExtended(IConsole::IResult*, void*)’:
src/engine/server/server.cpp:2466:46: error: expected ‘)’ before ‘;’ token
      pThis->GameServer()->GetClientVersion(i);
                                              ^
src/engine/server/server.cpp:2467:5: error: expected primary-expression before ‘)’ token
     );
     ^
bam: 'objs/engine/server/server_sql.o' error 256
```

**Comment by bretonium** [comment]
seems to work, thank you!

**Comment by Kaffeine** [comment]
After all this git-kung-fu the PR merged without squash? RIP `git bisect`, I'm used to use you here.

**Comment by bretonium** [comment]
> After all this git-kung-fu the PR merged without squash? RIP `git bisect`, I'm used to use you here.

how will 11-long commit history in a subtree affect `git bisect`?

**Comment by Kaffeine** [comment]
For `bisect` git performs a binary search. If you tell git that there is a good commit before this branch merged and a bad commit after this branch merged then there is a certain probability to have a commit from this branch be picked as a subject for bisection test and the test will obviously fail. 💁‍♂️
You can add 100 more commits after this branch merged and yet if you do `git bisect` against something like 3b14b85 (which is about 103 commits before this merged) then you'll hit one of those broken commits.
Yes, git have a smart graph traversing algorithms so it will consider merges properly. But until you tell it that you have the same result (bad/bad or good/good) on both ends of the branch fork and merge points it's free to pick one of those `Fix error ";"` commits.

**Comment by bretonium** [comment]
> there is a certain probability to have a commit from this branch be picked as a subject for bisection test and the test will obviously fail.

Although this is true if you _simply run_ `git bisect`, it can be easily avoided, if you know how to use `git bisect skip` or how to return code 125. Both of which are described in `man git-bisect` together with basic examples of how to use it.

**Comment by Kaffeine** [comment]
> it can be easily avoided

Your 'easy' way look like this:
1. during bisect git checks out commit
```e2691555b Merge branch 'Feature-ban-versions' of https://github.com/BloodWod```
2. The developer reaction:
"a-ha! Of course everybody know that this `e269155` commit is broken! Let's skip it".
or:
"a-ha! The message contains 'BloodWod' so probably the commit is broken! Let's skip it!"

While it is "easy" to avoid the issue, it could be avoided completely by using appropriate merge type.

**Comment by bretonium** [comment]
> The developer reaction:
>  The "easy" way implies that those broken commits are "well-known" for the developer.

The developer should stop reacting and start finally using automation tools (and error 125) built **specifically for this case** (as pointed out in all manuals and examples)

**Comment by Kaffeine** [comment]
How a developer can add those commits to the exception list (to raise code 125)?
Do you have such skip-list in the repository maybe? I'm familiar with bisect automation but I don't see how `git bisect run <script>` can skip the commits from this PR in a developer-friendly manner. I suggested two options: you either remember the hashes or skip commits from BloodWod.

My point: this merge made it difficult to use git bisect. `code 125` doesn't magically make it easier (it is a solution for a different problem).

There is no issue with skipping commits. There is an issue with "skip commits from this PR". You need some source of data for automation tools to omit these commits.

**Comment by bretonium** [comment]
> How a developer can add those commits to the exception list (to raise code 125)?

bam server_sql_release || exit 125

**Comment by Kaffeine** [comment]
It is way too late. You're suggesting to run the build. It is pointless to do `skip` at this point.

**Pull request merged**